core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2179,12 +2164,12 @@ private[spark] class DAGScheduler(
           val message = s"Stage failed because barrier task $task finished unsuccessfully.\n" +
             failure.toErrorString
           try {
-            // killAllTaskAttempts will fail if a SchedulerBackend does not implement killTask.
+            // cancelTasks will fail if a SchedulerBackend does not implement killTask.
             val reason = s"Task $task from barrier stage $failedStage (${failedStage.name}) " +
               "failed."
             val job = jobIdToActiveJob.get(failedStage.firstJobId)
             val shouldInterrupt = job.exists(j => shouldInterruptTaskThread(j))
-            taskScheduler.killAllTaskAttempts(stageId, shouldInterrupt, reason)
```

Comment:
I think that there is a subtle difference here: `killAllTaskAttempts` only kills tasks, whereas `cancelTasks` also calls `tsm.abort()` on the stage attempts, which might enqueue a new `taskSetFailed` event for each task set and I think that could have unintended side effects. Can you double-check whether we think that change is okay?

Reply:
This's really a good point. `taskSetFailed` will abort the stage and in turn fails the whole job, which is not the intended  behaviour here. The problem of `killAllTaskAttempts` is that it doesn't mark the `TaskSetManager` as zombie after killing all the tasks. So the `TaskSetManager` could still launch new tasks (by retry), which is not expected.

But I'm also thinking do we really want to abort the stages in `cancelTasks`? `cancelTasks` is currently called inside `cancelRunningIndependentStages` only. And `cancelRunningIndependentStages` is directly or indirectly called in 3 cases:

* When a job successfully finished: in this case, we expect that all the stages in this job can release the computation resources (i.e., kill all the tasks via `cleanupStateForJobAndIndependentStages`) immediately. But I think we don't expect this "release" action would lead to the stage abortion and in turn fail the job in the end. It doesn't fail the already succeeded job today because the succeeded job has been clean up (no longer exists in the `activeJobs` list) when the `taskSetFailed` event comes.
 
* When a job is requested to cancel: this case is essentially the same with the above case but only the job finishes in different states.

* When a stage aborts: in this case, we expect all the active jobs which depends on this stage to be canceled. Thus, we need to call `cancelRunningIndependentStages` on each active job. And this would finally fallback to the first case as the active job will be cleaned up (via `cleanupStateForJobAndIndependentStages`) first before the  `taskSetFailed` event comes.

Reply:
I agree. The worst case is we manually trigger "abort stage" in the existing callers of `cancelTasks` to keep the old behavior.

Reply:
Updated. I removed the "abort stage" inside `cancelTasks` but move it to after the callers of `cancelTasks`.  I also added a new conf `spark.scheduler.stage.abortStageAfterCancelTasks` to control wether we should abort the stage after `cancelTasks`. By default, we don't abort.